### PR TITLE
be:spdk: process spdk events before reusing controller

### PR DESCRIPTION
**Pull request by LiadOz**
Signed-off-by: Liad Oz <liadozil@gmail.com>

### CODE REVIEW

**Comment by LiadOz** [comment]
resolves #109

**Comment by safl** [comment]
Hi @LiadOz,
LGTM, CI is spinning up to check for any regressions.
Thanks,
Simon

**Comment by safl** [comment]
Looks like the test ``xnvme_tests_enum open --count 4`` fails with this change.
The test is here ``tests/enum.c``. In the failing scenario it attempts to open the same PCIe address multiple times, this is the use-case the ``cref`` was made for, as we cannot have multiple driver instances, we have to share and re-use the one we have previously attached to. I am currently traveling, so can't dive into the code, however @birkelund and @Baekalfen did you see a similar thing / issues when references counting in https://github.com/OpenMPDK/xNVMe/pull/111 ? Something that might be applicable here?

**Comment by birkelund** [comment]
> Looks like the test `xnvme_tests_enum open --count 4` fails with this change. The test is here `tests/enum.c`. In the failing scenario it attempts to open the same PCIe address multiple times, this is the use-case the `cref` was made for, as we cannot have multiple driver instances, we have to share and re-use the one we have previously attached to. I am currently traveling, so can't dive into the code, however @birkelund and @Baekalfen did you see a similar thing / issues when references counting in #111 ? Something that might be applicable here?

Test is OK for the be-vfio backend.

**Comment by LiadOz** [comment]
Is there a way to see the logs for this test? I suspect the controller might be detached when using it similar to https://github.com/OpenMPDK/xNVMe/issues/107, before this change assuming the controller was detached it gets to spdk_nvme_ctrlr_get_ns which returns 0 and then creates a new controller.

File: lib/xnvme_be_spdk_dev.c [hunk 1]
```diff
@@ -763,6 +763,16 @@ xnvme_be_spdk_state_init(struct xnvme_dev *dev)
 
 		XNVME_DEBUG("INFO: found dev->ident.uri: '%s' via cref_lookup()", dev->ident.uri);
 
+		err = spdk_nvme_ctrlr_process_admin_completions(state->ctrlr);
+		if (err < 0) {
+			XNVME_DEBUG("FAILED: spdk_nvme_ctrlr_process_admin_completions, err: %d",
+				    err);
+			if (_cref_deref(state->ctrlr)) {
```
Comment (karlowich) [inline lib/xnvme_be_spdk_dev.c:770]: @safl @LiadOz This line needs to be changed to:
`if (_cref_deref(state->ctrlr, true)) {`

**Review by karlowich — APPROVED** [review]
LGTM!

**Review by safl — APPROVED** [review]
Great, thanks!

**Pull request merged**